### ISSUE TITLE
query parameters in service endpoint are ignored

### ISSUE THREAD

**Issue by pomakis** [opengeospatial/ets-wcs20#32]
Version 1.11 of the CITE test suite for WCS 2.0.1 requires a service endpoint to be "specified without any query parameters".  If any query parameters are supplied, they're ignored.

However, nowhere in the WCS 2.0 core specification (OGC 09-110r4) does restriction exist.  In fact, section 11.2 of the OWS Common 2.0 specification (OGC 06-121r9), on which WCS 2.0 is based, states the following:

    An Online Resource URL intended for HTTP GET requests is in fact
    only a URL prefix to which additional parameters are appended in
    order to construct a valid operation request. The prefix defines
    the network address to which request messages are to be sent
    for a particular operation on a particular server, and may also
    identify a configuration of that server. Each operation may have
    a different prefix, and each prefix is entirely at the discretion
    of the service provider.
    
    A URL prefix is defined as a string including, in order, the scheme
    ("http" or "https"), Internet Protocol hostname or numeric address,
    optional port number, path, mandatory question mark '?', and optional
    string comprising one or more server-specific parameters ending in
    a mandatory ampersand '&'. Thus, an HTTP GET URL prefix always ends
    in either a question mark '?' or an ampersand '&'.

In our experience, it's quite common in the industry to provide one or more vendor-specific parameters in the URL prefix.  For example, the URL prefixes of CubeWerx's WCS servers have a required "DATASTORE" parameter that indicate a data set.  Such parameters are opaque to WCS clients, as long as the correctly follow the rules that the OWS Common 2.0 specification prescribes for appending HTTP GET parameters to the URL prefix declared for the appropriate operation.

E.g., I'm trying to test the service at:

https://demo.cubewerx.com/cubewerx/cubeserv/cite_wcs201?DATASTORE=Foundation&

but I can't, because the DATASTORE=Foundation parameter in the URL I supply is ignored, causing the very first test to fail.

As an aside, the WCS 2.0 core specification doesn't even use the term "service endpoint".  The common interpretation of "service endpoint" for OWS-Common-based services is the URL prefix that the service declares as the online resource for the GetCapabilities operation by the GetCapabilities document.  That's a bit of a mess, but these services were defined before OGC's more recent thrust to make its services more RESTful in nature.  So it is what it is.

**Comment by pebau** [comment]
Keith-
by way of normative inclusion, OWS Common rules are also applicable - WCS does not repeat these. Hence, ignoring extra parameters is the expected behavior.
THH,
Peter

PS: by way of history, OWS Common 2.0 was released after WCS 2 - we were forced, however, to change to it during the WCS adoption process. Today everybody knows that OWS Common 2 is a mess, and I do hope we get something better in due course.

**Comment by pomakis** [comment]
I disagree.  The above OWS Common excerpt clearly indicates that WCS requests need to include any query parameters that are part of the URL prefix.

**Comment by lgoltz** [comment]
I've tried with the follwing services: https://demo.cubewerx.com/cubewerx/cubeserv/cite_wcs201?DATASTORE=Foundation&SERVICE=WCS&REQUEST=GetCapabilities&ACCEPTVERSIONS=2.0.1

Test `wcs2:main_test` fails with the following log:
```
Test wcs2:main_test type Mandatory default result Passed (s0022/d675e37_1)

Assertion: Test the WCS 2.0.

Request d675e183_1:
   Method: get
   URL: https://demo.cubewerx.com/cubewerx/cubeserv/cite_wcs201?SERVICE=WCS&VERSION=2.0.1&REQUEST=GetCapabilities
   Response from parser ctlp:XMLValidatingParser:
      <ExceptionReport xmlns="http://www.opengis.net/ows/2.0"
                 xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                 version="2.0.1"
                 xml:lang="en"
                 xsi:schemaLocation="http://www.opengis.net/ows/2.0 http://schemas.cubewerx.com/schemas/ows/2.0.0/owsExceptionReport.xsd">
  <Exception exceptionCode="MissingParameterValue" locator="DATASTORE">
      <ExceptionText>CubeSERV: WCS GetCapabilities operation failed, raised in Wcs_Execute() in file "wcs.c" line 236 (process 86969)</ExceptionText>
      <ExceptionText>CubeSERV: Wcs_ParseCommonCgiParams() failed</ExceptionText>
      <ExceptionText>CubeSERV: Missing parameter "DATASTORE", raised in throwMissingParamError() in file "cw_cgiparams.c" line 1113 (process 86969)</ExceptionText>
  </Exception>
   <!--Response received in [13815] milliseconds--></ExceptionReport>

Message d675e218_1:
   [MESSAGE] Test the core and extensions when they are available.

Message d675e233_1:
   [FAILURE] The service does not provide at least one of the required protocol binding extensions.

Result: Failed
```

I also understand from OWS Common 2.0 that additional query parameters like `DATASTORE=Foundation` are allowed.

**Comment by keshav-nangare** [comment]
@dstenger 

Following is the way we are going to fix this extra parameter ignore issue:

1. The tests are not using the query parameter from the provided service. In every test, service is appended with SERVICE, REQUEST & ACCEPTVERSIONS parameter.
2. We will do http ctl:request and will parse the response and will get the URL from the GetCapabilities OperationMetadata with GET/href. This URL has the extra parameter i.e. DATASTORE.  
3. We will pass this URL to each of conformance class for further execution of the subtests.
4. As URL itself has the extra parameter and remaining will be attached by the test so the test is able to make http request successfully.

**Comment by keshav-nangare** [comment]
Fix with #95 PR.
